actions/rules/action_state_to_metastate.go

```diff
@@ -0,0 +1,467 @@
+			return errs.Wrap(err, "error updating work item")
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return workitem, nil
+}
+
+func (act ActionStateToMetaState) getValueListFromEnumField(wit *workitem.WorkItemType, fieldName string) ([]interface{}, error) {
+	enumFieldType := wit.Fields[fieldName].Type
+	switch t := enumFieldType.(type) {
+	case workitem.EnumType:
+		return t.Values, nil
+	}
+	return nil, errs.New("given field on workitemtype " + wit.ID.String() + " is not an enum field: " + fieldName)
+}
+
+// getStateToMetastateMap returns the mapping from state to metastate values read from the template.
+func (act ActionStateToMetaState) getStateToMetastateMap(workitemTypeID uuid.UUID) (map[string]string, error) {
+	wit, err := act.Db.WorkItemTypes().Load(act.Ctx, workitemTypeID)
+	if err != nil {
+		return nil, err
+	}
+	stateList, err := act.getValueListFromEnumField(wit, workitem.SystemState)
+	if err != nil {
+		return nil, err
+	}
+	metastateList, err := act.getValueListFromEnumField(wit, workitem.SystemMetaState)
+	if err != nil {
+		return nil, err
+	}
+	if len(stateList) != len(metastateList) {
+		return nil, errs.Errorf("inconsistent number of states and metatstates in the current work item type (must be equal): %d states != %d metastates", len(stateList), len(metastateList))
+	}
+	stateToMetastateMap := make(map[string]string)
+	for idx := range stateList {
+		thisState, ok := stateList[idx].(string)
+		if !ok {
+			return nil, errs.New("state value in value list is not of type string")
+		}
+		thisMetastate, ok := metastateList[idx].(string)
+		if !ok {
+			return nil, errs.New("metastate value in value list is not of type string")
+		}
+		// only the first state<->metatstate mapping needs to be taken as
+		// per definition. So we're only adding a new entry here if there
+		// is not an entry already existing.
+		if _, ok := stateToMetastateMap[thisState]; !ok {
+			stateToMetastateMap[thisState] = thisMetastate
+		}
+	}
+	return stateToMetastateMap, nil
+}
+
+// getMetastateToStateMap returns the mapping from metastate to state values read from the template.
+func (act ActionStateToMetaState) getMetastateToStateMap(workitemTypeID uuid.UUID) (map[string]string, error) {
+	stateToMetastate, err := act.getStateToMetastateMap(workitemTypeID)
+	if err != nil {
+		return nil, err
+	}
+	metastateToStateMap := make(map[string]string)
+	for state, metastate := range stateToMetastate {
+		// this is important: we only add a new entry if there
+		// is not an entry already existing. Therefore, we're only
+		// using the first mapping, satisfying the req for getting the
+		// first matching state for a metastate.
+		if _, ok := metastateToStateMap[metastate]; !ok {
+			metastateToStateMap[metastate] = state
+		}
+	}
+	return metastateToStateMap, nil
+}
+
+func (act ActionStateToMetaState) addOrUpdateChange(changes change.Set, attributeName string, oldValue interface{}, newValue interface{}) change.Set {
```

Comment:
Why is this a member function when `act` is not used anywhere inside?

Reply:
Because there is no need to spill over purely internal functionality to package level. Encapsulation.

Reply:
I didn't say you should put it on the package level but how about making it part of the `change.Set`?

Reply:
Also, one could argue about the reason you gave for not putting it as a function on package level. Here's you may not be polluting the package level but you're requiring to create an object, namely `ActionStateToMetaState`, that is not needed for the execution of the function. I consider this worse because the prerequisites to run the function are just wrong.

Reply:
I think you missed some of the basic ideas here. There is good reason for having Actions encapsulated as structs.

Reply:
> I think you missed some of the basic ideas here. There is good reason for having Actions encapsulated as structs.

You misunderstood me. I never said to not store actions as structs. All I asked for was to outsource helper functions as testable-building blocks. We've discussed that in today's Bluejeans meeting. As much as I don't like it but I can live with a test that does `ActionStateToMetaState{}.addOrUpdateChange()`.

Reply:
Ok. Fixed in 7dafce4.